[PWPA 1996] Configure advanced options for SCORE converter

gabrielantao opened the pull request:
Configure some values for the physics, numerical, and time options. Now the builder classes keep the parsed data as attributes, to avoid call reader methods all the time in order to get operation data.

PWPA-1996

Commit:
Add configurations for physics;

PWPA-1996

Commit:
Use operation data as attribute of operation class

PWPA-1996

Commit:
Add numerical and time options;

PWPA-1996

Commit:
Define the hydrodynamic model;

PWPA-1996

Commit:
Fix the operation data missing attribute;

PWPA-1996

carlosnewmar commented on src/alfasim_score/units.py:
```diff
@@ -16,3 +16,4 @@
 HEAT_TRANSFER_COEFFICIENT_UNIT = "W/m2.K"
 DIMENSIONLESS = "unitless"
 VELOCITY_UNIT = "m/s"
+OPERATION_DURATION_UNIT = "min"
```
`TIME_UNIT` seems more consisntent with the rest of constants.

gabrielantao replied:
ok, done.

carlosnewmar commented on src/alfasim_score/converter/alfacase/production_operation.py:
```diff
@@ -117,14 +123,38 @@ def configure_well_initial_conditions(self, alfacase: CaseDescription) -> None:
+    def configure_physics(self, alfacase: CaseDescription) -> None:
+        """Configure the description for the physics data."""
+        super().configure_physics(alfacase)
+        no_inlet_water = np.isclose(self.general_data["water_flow_rate"].GetValue(), 0.0)
```
The same way we have `self.has_gas_lift()` we could have a `self.has_water()`.
It could check if `self.general_data["water_flow_rate"]` is zero (or close) for production operation and `fluid_type` for injection operation.

gabrielantao replied:
I created new methods to encapsulate these checks.

carlosnewmar commented on src/alfasim_score/converter/alfacase/base_operation.py:
```diff
@@ -154,14 +162,29 @@ def configure_well_initial_conditions(self, alfacase: CaseDescription) -> None:
         )
 
-    # TODO PWPA-1996: review the configurations done here.
     def configure_physics(self, alfacase: CaseDescription) -> None:
```
At a glance, it seems we don't need to overwrite this method.
```python
    def configure_physics(self, alfacase: CaseDescription) -> None:
        """Configure the description for the physics data."""
        alfacase.physics = PhysicsDescription(
            hydrodynamic_model= HydrodynamicModelType.FourFields if self.has_water() else HydrodynamicModelType.ThreeLayersGasOilWater,
            energy_model=EnergyModel.GlobalModel,
            simulation_regime=SimulationRegimeType.Transient if self.has_gas_lift() else SimulationRegimeType.SteadyState,
            initial_condition_strategy=InitialConditionStrategyType.Constant,
        )
```

But only `has_water` and `has_gas_lift`, thus the code will be much more simple.

gabrielantao replied:
One reason I splited these two methods is because I don't have the `has_gas_lift` in the base class because the data to build it is just available for the production operation file and I it will be weird in the base once I think it only makes sense for the `ProductionOperationBuilder`. However, I think I can simplify by removing the method from the `InjectionOperationBuilder` if I set it in the base like this

```python
    def configure_physics(self, alfacase: CaseDescription) -> None:
        """Configure the description for the physics data."""
        alfacase.physics = PhysicsDescription(
            hydrodynamic_model=HydrodynamicModelType.ThreeLayersGasOilWater,
            energy_model=EnergyModel.GlobalModel,
            # moved parameters from InjectionOperationBuilder
            simulation_regime=SimulationRegimeType.SteadyState,
            initial_condition_strategy=InitialConditionStrategyType.Constant,
        )
```
It feels a little bit simpler to me than the way it was implemented. What do you think?

gabrielantao replied:
Done.

carlosnewmar replied:
It is not weird `BaseOperationBuilder` has a `has_gas_lift` returning `False` and it be overwritten in `ProductionOperationBuilder` with the correct checks. The same for the `has_water`.
IMHO the code could be more clear and straight as I suggested. But it is up to you to decide it.

gabrielantao replied:
ok.

Commit:
Simplify physics configuration;

Change time unit name;
Add methods to check if there is water or gas in well;

PWPA-1996

carlosnewmar commented on src/alfasim_score/converter/alfacase/injection_operation.py:
```diff
@@ -20,60 +21,82 @@
+        has_inlet_flow = self.general_data["flow_rate"].GetValue() > 0.0
+        has_initial_water = np.any(
+            np.array(
+                alfacase.wells[0].initial_conditions.volume_fractions.table_length.fractions.get(
```
At the end of day, with this check we are always checking `self.general_data["fluid_type"]` twice, isn't?
Because, we set the initial fraction using `self.general_data["fluid_type"]`, thus is this method, we check the initial volume fraction (that used `self.general_data["fluid_type"]`).
Thus, in the next line we check `self.general_data["fluid_type"]` again.
I'm being repetitive, because it seems the code is repetitive 😃

gabrielantao replied:
Indeed it was duplicated. Changed to just check the type of fluid and it makes sure there is inlet flow.

carlosnewmar commented on src/alfasim_score/constants.py:
```diff
@@ -46,3 +46,10 @@
+# numerical and time options default values
+MAXIMUM_TIMESTEP_CHANGE_FACTOR = 1.2
+NUMERICAL_TOLERANCE = 1.0e-3
+INITIAL_TIMESTEP = Scalar(0.1, TIME_UNIT)
```
```
INITIAL_TIMESTEP = Scalar(0.1, TIME_UNIT)
MINIMUM_TIMESTEP = Scalar(1.0e-4, TIME_UNIT)
MAXIMUM_TIMESTEP = Scalar(2.0, TIME_UNIT)
```
These values are supposed to be in seconds, but `TIME_UNIT` is minutes.

gabrielantao replied:
Fixed.

Commit:
Remove duplication in check of fluid;

Fix the unit of the timestep;

PWPA-1996
